File: opendevin/wire.py
"""Minimal protobuf wire codec (pure Python, no protobuf dependency).

Implements just the subset of the Cognition `exa.*` schemas that OpenDevin
needs, with field numbers taken from the official descriptors shipped in the
`jeopi-catalog` npm package (see `proto/*.fdp` in this repo) and verified
against live traffic captured from `devin 3000.10.21`.

Only varint, length-delimited (string/bytes/message) and fixed64/fixed32
wire types are used by the messages we touch.
"""
from __future__ import annotations

import struct
from typing import Any, Dict, List, Optional, Tuple


def _varint(n: int) -> bytes:
    n &= 0xFFFFFFFFFFFFFFFF
    out = bytearray()
    while True:
        b = n & 0x7F
        n >>= 7
        if n:
            out.append(b | 0x80)
        else:
            out.append(b)
            return bytes(out)


def _tag(field: int, wire: int) -> bytes:
    return _varint((field << 3) | wire)


def _len(v: bytes) -> bytes:
    return _varint(len(v)) + v


def _field_str(field: int, s: str) -> bytes:
    return _tag(field, 2) + _len(s.encode("utf-8"))


def _field_bytes(field: int, b: bytes) -> bytes:
    return _tag(field, 2) + _len(b)


def _field_msg(field: int, m: bytes) -> bytes:
    return _field_bytes(field, m)


def _field_varint(field: int, n: int) -> bytes:
    return _tag(field, 0) + _varint(n)


def _field_f64(field: int, x: float) -> bytes:
    return _tag(field, 1) + struct.pack("<d", x)


def _field_f32(field: int, x: float) -> bytes:
    return _tag(field, 5) + struct.pack("<f", x)


def _field_bool(field: int, b: bool) -> bytes:
    return _field_varint(field, 1 if b else 0)


def _msg(fields: bytes) -> bytes:
    return fields


# ---------------------------------------------------------------------------
# Encode helpers for the messages we build
# ---------------------------------------------------------------------------

def encode_chat_tool_call(tc: Dict[str, str]) -> bytes:
    out = b""
    if tc.get("id"):
        out += _field_str(1, tc["id"])
    if tc.get("name"):
        out += _field_str(2, tc["name"])
    if tc.get("arguments_json"):
        out += _field_str(3, tc["arguments_json"])
    return _msg(out)


def encode_image(data: Dict[str, str]) -> bytes:
    out = b""
    if data.get("base64_data"):
        out += _field_str(1, data["base64_data"])
    if data.get("mime_type"):
        out += _field_str(2, data["mime_type"])
    return _msg(out)


def encode_chat_message_prompt(p: Dict[str, Any]) -> bytes:
    out = b""
    if p.get("message_id"):
        out += _field_str(1, p["message_id"])
    if p.get("source"):
        out += _field_varint(2, p["source"])
    if p.get("prompt"):
        out += _field_str(3, p["prompt"])
    for tc in p.get("tool_calls") or []:
        out += _field_msg(6, encode_chat_tool_call(tc))
    if p.get("tool_call_id"):
        out += _field_str(7, p["tool_call_id"])
    for img in p.get("images") or []:
        out += _field_msg(10, encode_image(img))
    if p.get("thinking"):
        out += _field_str(11, p["thinking"])
    if p.get("signature"):
        out += _field_str(12, p["signature"])
    if p.get("thinking_id"):
        out += _field_str(16, p["thinking_id"])
    if p.get("signature_type"):
        out += _field_str(18, p["signature_type"])
    if p.get("phase"):
        out += _field_str(19, p["phase"])
    return _msg(out)


def encode_completion_config(conf: Dict[str, Any]) -> bytes:
    out = b""
    if "num_completions" in conf:
        out += _field_varint(1, conf["num_completions"])
    if "max_tokens" in conf:
        out += _field_varint(2, conf["max_tokens"])
    if "max_newlines" in conf:
        out += _field_varint(3, conf["max_newlines"])
    if "temperature" in conf:
        out += _field_f64(5, conf["temperature"])
    if "top_k" in conf:
        out += _field_varint(7, conf["top_k"])
    if "top_p" in conf:
        out += _field_f64(8, conf["top_p"])
    if "first_temperature" in conf:
        out += _field_f64(9, conf["first_temperature"])
    if "fim_eot_prob_threshold" in conf:
        out += _field_f64(10, conf["fim_eot_prob_threshold"])
    return _msg(out)


def encode_tool_definition(t: Dict[str, Any]) -> bytes:
    out = b""
    if t.get("name"):
        out += _field_str(1, t["name"])
    if t.get("description"):
        out += _field_str(2, t["description"])
    if t.get("json_schema_string"):
        out += _field_str(3, t["json_schema_string"])
    if "strict" in t and t["strict"]:
        out += _field_bool(4, t["strict"])
    return _msg(out)


def encode_metadata(md: Dict[str, Any]) -> bytes:
    out = b""
    if md.get("ide_name"):
        out += _field_str(1, md["ide_name"])
    if md.get("extension_version"):
        out += _field_str(2, md["extension_version"])
    if md.get("api_key"):
        out += _field_str(3, md["api_key"])
    if md.get("locale"):
        out += _field_str(4, md["locale"])
    if md.get("os"):
        out += _field_str(5, md["os"])
    if md.get("ide_version"):
        out += _field_str(7, md["ide_version"])
    if md.get("request_id"):
        out += _field_varint(9, md["request_id"])
    if md.get("session_id"):
        out += _field_str(10, md["session_id"])
    if md.get("extension_name"):
        out += _field_str(12, md["extension_name"])
    if md.get("ide_type"):
        out += _field_str(28, md["ide_type"])
    if md.get("identity_digest"):
        out += _field_str(31, md["identity_digest"])
    return _msg(out)


def encode_get_chat_message_request(r: Dict[str, Any]) -> bytes:
    out = b""
    if r.get("metadata"):
        out += _field_msg(1, encode_metadata(r["metadata"]))
    if r.get("prompt"):
        out += _field_str(2, r["prompt"])
    for p in r.get("chat_message_prompts") or []:
        out += _field_msg(3, encode_chat_message_prompt(p))
    if r.get("request_type"):
        out += _field_varint(7, r["request_type"])
    if r.get("configuration"):
        out += _field_msg(8, encode_completion_config(r["configuration"]))
    for t in r.get("tools") or []:
        out += _field_msg(10, encode_tool_definition(t))
    if r.get("cascade_id"):
        out += _field_str(16, r["cascade_id"])
    if r.get("planner_mode"):
        out += _field_varint(20, r["planner_mode"])
    if r.get("chat_model_uid"):
        out += _field_str(21, r["chat_model_uid"])
    if r.get("execution_id"):
        out += _field_str(22, r["execution_id"])
    return _msg(out)


# ---------------------------------------------------------------------------
# Decode helpers for the response stream
# ---------------------------------------------------------------------------

def _read_varint(b: bytes, i: int) -> Tuple[int, int]:
    r = 0
    s = 0
    while True:
        x = b[i]
        i += 1
        r |= (x & 0x7F) << s
        if not x & 0x80:
            return r, i
        s += 7


def iter_fields(b: bytes):
    """Yield (field_number, wire_type, raw_value) for a protobuf message."""
    i = 0
    n = len(b)
    while i < n:
        tag, i = _read_varint(b, i)
        fno, wt = tag >> 3, tag & 7
        if wt == 0:
            v, i = _read_varint(b, i)
            yield fno, wt, v
        elif wt == 1:
            yield fno, wt, b[i:i + 8]
            i += 8
        elif wt == 2:
            ln, i = _read_varint(b, i)
            yield fno, wt, b[i:i + ln]
            i += ln
        elif wt == 5:
            yield fno, wt, b[i:i + 4]
            i += 4
        else:
            break


class ToolCall:
    __slots__ = ("id", "name", "arguments_json")

    def __init__(self):
        self.id = ""
        self.name = ""
        self.arguments_json = ""


class Usage:
    __slots__ = ("input_tokens", "output_tokens", "cache_read_tokens",
                 "cache_write_tokens", "model_uid", "api_provider")

    def __init__(self):
        self.input_tokens = 0
        self.output_tokens = 0
        self.cache_read_tokens = 0
        self.cache_write_tokens = 0
        self.model_uid = ""
        self.api_provider = 0


class ChatMessageResponse:
    __slots__ = ("message_id", "delta_text", "delta_tokens", "stop_reason",
                 "delta_thinking", "delta_signature", "delta_signature_type",
                 "thinking_id", "phase", "latency", "request_id",
                 "usage", "tool_calls")

    def __init__(self):
        self.message_id = ""
        self.delta_text = ""
        self.delta_tokens = 0
        self.stop_reason = 0
        self.delta_thinking = ""
        self.delta_signature = ""
        self.delta_signature_type = ""
        self.thinking_id = ""
        self.phase = ""
        self.latency = 0.0
        self.request_id = ""
        self.usage = None  # Usage
        self.tool_calls = []  # list[ToolCall]


def parse_tool_call(b: bytes) -> ToolCall:
    tc = ToolCall()
    for fno, wt, v in iter_fields(b):
        if fno == 1 and wt == 2:
            tc.id = v.decode("utf-8", "replace")
        elif fno == 2 and wt == 2:
            tc.name = v.decode("utf-8", "replace")
        elif fno == 3 and wt == 2:
            tc.arguments_json = v.decode("utf-8", "replace")
    return tc


def parse_usage(b: bytes) -> Usage:
    u = Usage()
    for fno, wt, v in iter_fields(b):
        if fno == 2 and wt == 0:
            u.input_tokens = v
        elif fno == 3 and wt == 0:
            u.output_tokens = v
        elif fno == 4 and wt == 0:
            u.cache_write_tokens = v
        elif fno == 5 and wt == 0:
            u.cache_read_tokens = v
        elif fno == 9 and wt == 2:
            u.model_uid = v.decode("utf-8", "replace")
        elif fno == 6 and wt == 0:
            u.api_provider = v
    return u


def parse_chat_message_response(b: bytes) -> ChatMessageResponse:
    r = ChatMessageResponse()
    for fno, wt, v in iter_fields(b):
        if fno == 1 and wt == 2:
            r.message_id = v.decode("utf-8", "replace")
        elif fno == 3 and wt == 2:
            r.delta_text = v.decode("utf-8", "replace")
        elif fno == 4 and wt == 0:
            r.delta_tokens = v
        elif fno == 5 and wt == 0:
            r.stop_reason = v
        elif fno == 6 and wt == 2:
            r.tool_calls.append(parse_tool_call(v))
        elif fno == 7 and wt == 2:
            r.usage = parse_usage(v)
        elif fno == 9 and wt == 2:
            r.delta_thinking = v.decode("utf-8", "replace")
        elif fno == 10 and wt == 2:
            r.delta_signature = v.decode("utf-8", "replace")
        elif fno == 12 and wt == 1:
            r.latency = struct.unpack("<d", v)[0]
        elif fno == 16 and wt == 2:
            r.thinking_id = v.decode("utf-8", "replace")
        elif fno == 17 and wt == 2:
            r.request_id = v.decode("utf-8", "replace")
        elif fno == 21 and wt == 2:
            r.delta_signature_type = v.decode("utf-8", "replace")
        elif fno == 25 and wt == 2:
            r.phase = v.decode("utf-8", "replace")
    return r

File: source/opendevin-cli/src/wire.rs
//! Minimal protobuf wire codec for the Cognition `exa.*` schemas used by
//! OpenDevin. Field numbers are taken from the official descriptors shipped in
//! the `jeopi-catalog` npm package and verified against live traffic captured
//! from the real `devin 3000.10.21` binary (see docs/Protocol.md).

pub fn varint(n: u64) -> Vec<u8> {
    let mut out = Vec::new();
    let mut v = n;
    loop {
        let b = (v & 0x7f) as u8;
        v >>= 7;
        if v != 0 {
            out.push(b | 0x80);
        } else {
            out.push(b);
            return out;
        }
    }
}

pub fn tag(field: u32, wire: u32) -> Vec<u8> {
    varint(((field as u64) << 3) | wire as u64)
}

fn len_prefixed(b: &[u8]) -> Vec<u8> {
    let mut out = varint(b.len() as u64);
    out.extend_from_slice(b);
    out
}

pub fn field_str(field: u32, s: &str) -> Vec<u8> {
    let mut out = tag(field, 2);
    out.extend_from_slice(&len_prefixed(s.as_bytes()));
    out
}

pub fn field_bytes(field: u32, b: &[u8]) -> Vec<u8> {
    let mut out = tag(field, 2);
    out.extend_from_slice(&len_prefixed(b));
    out
}

pub fn field_msg(field: u32, m: &[u8]) -> Vec<u8> {
    field_bytes(field, m)
}

pub fn field_varint(field: u32, n: u64) -> Vec<u8> {
    let mut out = tag(field, 0);
    out.extend_from_slice(&varint(n));
    out
}

pub fn field_f64(field: u32, x: f64) -> Vec<u8> {
    let mut out = tag(field, 1);
    out.extend_from_slice(&x.to_le_bytes());
    out
}

pub fn field_bool(field: u32, b: bool) -> Vec<u8> {
    field_varint(field, if b { 1 } else { 0 })
}

// ---------------------------------------------------------------------------
// Encode helpers
// ---------------------------------------------------------------------------

pub fn encode_metadata(md: &Metadata) -> Vec<u8> {
    let mut out = Vec::new();
    if !md.ide_name.is_empty() {
        out.extend(field_str(1, &md.ide_name));
    }
    if !md.extension_version.is_empty() {
        out.extend(field_str(2, &md.extension_version));
    }
    if !md.api_key.is_empty() {
        out.extend(field_str(3, &md.api_key));
    }
    if !md.locale.is_empty() {
        out.extend(field_str(4, &md.locale));
    }
    if !md.os.is_empty() {
        out.extend(field_str(5, &md.os));
    }
    if !md.ide_version.is_empty() {
        out.extend(field_str(7, &md.ide_version));
    }
    if md.request_id != 0 {
        out.extend(field_varint(9, md.request_id));
    }
    if !md.session_id.is_empty() {
        out.extend(field_str(10, &md.session_id));
    }
    if !md.extension_name.is_empty() {
        out.extend(field_str(12, &md.extension_name));
    }
    if !md.ide_type.is_empty() {
        out.extend(field_str(28, &md.ide_type));
    }
    out
}

pub fn encode_chat_tool_call(tc: &ToolCall) -> Vec<u8> {
    let mut out = Vec::new();
    if !tc.id.is_empty() {
        out.extend(field_str(1, &tc.id));
    }
    if !tc.name.is_empty() {
        out.extend(field_str(2, &tc.name));
    }
    if !tc.arguments_json.is_empty() {
        out.extend(field_str(3, &tc.arguments_json));
    }
    out
}

pub fn encode_chat_message_prompt(p: &ChatMessagePrompt) -> Vec<u8> {
    let mut out = Vec::new();
    if !p.message_id.is_empty() {
        out.extend(field_str(1, &p.message_id));
    }
    if p.source != 0 {
        out.extend(field_varint(2, p.source as u64));
    }
    if !p.prompt.is_empty() {
        out.extend(field_str(3, &p.prompt));
    }
    for tc in &p.tool_calls {
        out.extend(field_msg(6, &encode_chat_tool_call(tc)));
    }
    if !p.tool_call_id.is_empty() {
        out.extend(field_str(7, &p.tool_call_id));
    }
    for img in &p.images {
        let mut im = Vec::new();
        if !img.base64_data.is_empty() {
            im.extend(field_str(1, &img.base64_data));
        }
        if !img.mime_type.is_empty() {
            im.extend(field_str(2, &img.mime_type));
        }
        out.extend(field_msg(10, &im));
    }
    if !p.thinking.is_empty() {
        out.extend(field_str(11, &p.thinking));
    }
    if !p.signature.is_empty() {
        out.extend(field_str(12, &p.signature));
    }
    out
}

pub fn encode_completion_config(c: &CompletionConfig) -> Vec<u8> {
    let mut out = Vec::new();
    if c.num_completions != 0 {
        out.extend(field_varint(1, c.num_completions as u64));
    }
    if c.max_tokens != 0 {
        out.extend(field_varint(2, c.max_tokens as u64));
    }
    if c.max_newlines != 0 {
        out.extend(field_varint(3, c.max_newlines as u64));
    }
    if (c.temperature - 0.0).abs() > f64::EPSILON {
        out.extend(field_f64(5, c.temperature));
    }
    if c.top_k != 0 {
        out.extend(field_varint(7, c.top_k as u64));
    }
    if (c.top_p - 0.0).abs() > f64::EPSILON {
        out.extend(field_f64(8, c.top_p));
    }
    if (c.first_temperature - 0.0).abs() > f64::EPSILON {
        out.extend(field_f64(9, c.first_temperature));
    }
    out
}

pub fn encode_tool_definition(t: &ToolDefinition) -> Vec<u8> {
    let mut out = Vec::new();
    if !t.name.is_empty() {
        out.extend(field_str(1, &t.name));
    }
    if !t.description.is_empty() {
        out.extend(field_str(2, &t.description));
    }
    if !t.json_schema_string.is_empty() {
        out.extend(field_str(3, &t.json_schema_string));
    }
    if t.strict {
        out.extend(field_bool(4, true));
    }
    out
}

pub fn encode_get_chat_message_request(r: &ChatRequest) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend(field_msg(1, &encode_metadata(&r.metadata)));
    if !r.prompt.is_empty() {
        out.extend(field_str(2, &r.prompt));
    }
    for p in &r.chat_message_prompts {
        out.extend(field_msg(3, &encode_chat_message_prompt(p)));
    }
    if r.request_type != 0 {
        out.extend(field_varint(7, r.request_type as u64));
    }
    out.extend(field_msg(8, &encode_completion_config(&r.configuration)));
    for t in &r.tools {
        out.extend(field_msg(10, &encode_tool_definition(t)));
    }
    if !r.cascade_id.is_empty() {
        out.extend(field_str(16, &r.cascade_id));
    }
    if r.planner_mode != 0 {
        out.extend(field_varint(20, r.planner_mode as u64));
    }
    if !r.chat_model_uid.is_empty() {
        out.extend(field_str(21, &r.chat_model_uid));
    }
    if !r.execution_id.is_empty() {
        out.extend(field_str(22, &r.execution_id));
    }
    out
}

// ---------------------------------------------------------------------------
// Decode helpers
// ---------------------------------------------------------------------------

pub struct Field {
    pub field: u32,
    pub wire: u32,
    pub value: FieldValue,
}

pub enum FieldValue {
    Varint(u64),
    Fixed64([u8; 8]),
    Fixed32([u8; 4]),
    Bytes(Vec<u8>),
}

pub fn read_varint(b: &[u8], i: &mut usize) -> Option<u64> {
    let mut r = 0u64;
    let mut s = 0u32;
    loop {
        let x = *b.get(*i)?;
        *i += 1;
        r |= ((x & 0x7f) as u64) << s;
        if x & 0x80 == 0 {
            return Some(r);
        }
        s += 7;
        if s > 63 {
            return None;
        }
    }
}

pub fn iter_fields(b: &[u8]) -> Vec<Field> {
    let mut out = Vec::new();
    let mut i = 0usize;
    while i < b.len() {
        let Some(tag) = read_varint(b, &mut i) else { break };
        let field = (tag >> 3) as u32;
        let wire = (tag & 7) as u32;
        match wire {
            0 => {
                let Some(v) = read_varint(b, &mut i) else { break };
                out.push(Field { field, wire, value: FieldValue::Varint(v) });
            }
            1 => {
                if i + 8 > b.len() {
                    break;
                }
                let mut arr = [0u8; 8];
                arr.copy_from_slice(&b[i..i + 8]);
                i += 8;
                out.push(Field { field, wire, value: FieldValue::Fixed64(arr) });
            }
            2 => {
                let Some(ln) = read_varint(b, &mut i) else { break };
                let ln = ln as usize;
                if i + ln > b.len() {
                    break;
                }
                let raw = b[i..i + ln].to_vec();
                i += ln;
                out.push(Field { field, wire, value: FieldValue::Bytes(raw) });
            }
            5 => {
                if i + 4 > b.len() {
                    break;
                }
                let mut arr = [0u8; 4];
                arr.copy_from_slice(&b[i..i + 4]);
                i += 4;
                out.push(Field { field, wire, value: FieldValue::Fixed32(arr) });
            }
            _ => break,
        }
    }
    out
}

// ---------------------------------------------------------------------------
// Message structs
// ---------------------------------------------------------------------------

#[derive(Default, Clone)]
pub struct Metadata {
    pub ide_name: String,
    pub ide_version: String,
    pub ide_type: String,
    pub extension_name: String,
    pub extension_version: String,
    pub api_key: String,
    pub locale: String,
    pub os: String,
    pub session_id: String,
    pub request_id: u64,
}

#[derive(Default, Clone)]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub arguments_json: String,
}

#[derive(Default, Clone)]
pub struct ImageData {
    pub base64_data: String,
    pub mime_type: String,
}

#[derive(Default, Clone)]
pub struct ChatMessagePrompt {
    pub message_id: String,
    pub source: u32,
    pub prompt: String,
    pub tool_calls: Vec<ToolCall>,
    pub tool_call_id: String,
    pub images: Vec<ImageData>,
    pub thinking: String,
    pub signature: String,
}

#[derive(Default, Clone)]
pub struct CompletionConfig {
    pub num_completions: u32,
    pub max_tokens: u32,
    pub max_newlines: u32,
    pub temperature: f64,
    pub top_k: u32,
    pub top_p: f64,
    pub first_temperature: f64,
}

#[derive(Default, Clone)]
pub struct ToolDefinition {
    pub name: String,
    pub description: String,
    pub json_schema_string: String,
    pub strict: bool,
}

#[derive(Default, Clone)]
pub struct ChatRequest {
    pub metadata: Metadata,
    pub prompt: String,
    pub chat_message_prompts: Vec<ChatMessagePrompt>,
    pub request_type: u32,
    pub configuration: CompletionConfig,
    pub tools: Vec<ToolDefinition>,
    pub cascade_id: String,
    pub planner_mode: u32,
    pub chat_model_uid: String,
    pub execution_id: String,
}

#[derive(Default, Clone)]
pub struct Usage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_tokens: u64,
    pub cache_write_tokens: u64,
    pub model_uid: String,
}

#[derive(Default, Clone)]
pub struct ChatMessageResponse {
    pub message_id: String,
    pub delta_text: String,
    pub delta_tokens: u32,
    pub stop_reason: u32,
    pub delta_thinking: String,
    pub delta_signature: String,
    pub thinking_id: String,
    pub phase: String,
    pub latency: f64,
    pub request_id: String,
    pub usage: Option<Usage>,
    pub tool_calls: Vec<ToolCall>,
}

pub fn parse_usage(b: &[u8]) -> Usage {
    let mut u = Usage::default();
    for f in iter_fields(b) {
        match f.field {
            2 => u.input_tokens = fv_varint(&f),
            3 => u.output_tokens = fv_varint(&f),
            4 => u.cache_write_tokens = fv_varint(&f),
            5 => u.cache_read_tokens = fv_varint(&f),
            9 => u.model_uid = fv_str(&f),
            _ => {}
        }
    }
    u
}

pub fn parse_tool_call(b: &[u8]) -> ToolCall {
    let mut tc = ToolCall::default();
    for f in iter_fields(b) {
        match f.field {
            1 => tc.id = fv_str(&f),
            2 => tc.name = fv_str(&f),
            3 => tc.arguments_json = fv_str(&f),
            _ => {}
        }
    }
    tc
}

pub fn parse_chat_message_response(b: &[u8]) -> ChatMessageResponse {
    let mut r = ChatMessageResponse::default();
    for f in iter_fields(b) {
        match f.field {
            1 => r.message_id = fv_str(&f),
            3 => r.delta_text = fv_str(&f),
            4 => r.delta_tokens = fv_varint(&f) as u32,
            5 => r.stop_reason = fv_varint(&f) as u32,
            6 => r.tool_calls.push(parse_tool_call(&fv_bytes(&f))),
            7 => r.usage = Some(parse_usage(&fv_bytes(&f))),
            9 => r.delta_thinking = fv_str(&f),
            10 => r.delta_signature = fv_str(&f),
            12 => r.latency = f64::from_le_bytes(fv_f64(&f)),
            16 => r.thinking_id = fv_str(&f),
            17 => r.request_id = fv_str(&f),
            25 => r.phase = fv_str(&f),
            _ => {}
        }
    }
    r
}

fn fv_varint(f: &Field) -> u64 {
    match &f.value {
        FieldValue::Varint(v) => *v,
        _ => 0,
    }
}

fn fv_str(f: &Field) -> String {
    match &f.value {
        FieldValue::Bytes(b) => String::from_utf8_lossy(b).into_owned(),
        _ => String::new(),
    }
}

fn fv_bytes(f: &Field) -> Vec<u8> {
    match &f.value {
        FieldValue::Bytes(b) => b.clone(),
        _ => Vec::new(),
    }
}

fn fv_f64(f: &Field) -> [u8; 8] {
    match &f.value {
        FieldValue::Fixed64(a) => *a,
        _ => [0u8; 8],
    }
}